contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
+of its controlled objects, as well as to let controllers wait for asynchronous
+operations to complete on those objects.
+The controller manager subscribes to a stream of events about controlled objects
+and routes each event to a particular controller.
+
+Previously, the controller manager used only label selectors to decide which
+controller to route an event to. If multiple controllers had overlapping
+selectors, events might be misrouted, causing the wrong controllers to sync,
+or causing some controllers to freeze because they keep waiting for an event
+that already happened.
+
+Some controller managers introduced a workaround that used the controller
+creation timestamps and names to break ties. However, that did not prevent
+misrouting if the overlapping controllers were of different types.
+It also only worked while controllers themselves assigned ownership over objects
+using the same tie-break rules.
+
+Now that controller ownership is defined in terms of ControllerRef,
+controller managers should use the following guidelines for responding to watch
+events:
+
+* If the object has a ControllerRef:
+  * Sync only the referenced controller.
+  * Update `expectations` counters for the referenced controller.
+  * If an *Update* event removes the ControllerRef, sync any controllers whose
+    selectors match to give each one a chance to adopt the object.
+* If the object is an orphan:
+  * *Add* event
+    * Sync any controllers whose selectors match to give each one a chance to
+      adopt the object.
+    * Do *not* update counters on `expectations`.
+      Controllers should never be waiting for creation of an orphan because
+      anything they create should have a ControllerRef.
+  * *Delete* event
+    * Do *not* sync any controllers.
+      Controllers should never care about orphans disappearing.
+    * Do *not* update counters on `expectations`.
+      Controllers should never be waiting for deletion of an orphan because they
+      are not allowed to delete objects they don't own.
+  * *Update* event
+    * If labels changed, sync any controllers whose selectors match to give each
+      one a chance to adopt the object.
+
+## Controller-specific behavior
+
+This section lists considerations specific to a given controller.
+
+* **StatefulSet**
+
+  * A StatefulSet will not adopt any Pod whose name does not match the template
+    it uses to create new Pods: `{statefulset name}-{ordinal}`.
+    This is because Pods in a given StatefulSet form a "family" that may use pod
+    names (via their generated DNS entries) to coordinate among themselves.
+    Adopting Pods with the wrong names would violate StatefulSet's semantics.
+
+    Adoption is allowed when Pod names match, so it remains possible to orphan a
+    family of Pods (by deleting their StatefulSet without cascading) and then
+    create a new StatefulSet with the same name and selector to adopt them.
```

Comment:
This doesn't seem to be the case as of 1.5.2. I just tried setting a controllerRef on a StatefulSet pod and non-cascading deletion of the controller blew the pod away as well.